src/shred_collector/shred_receiver.zig

```diff
@@ -0,0 +1,202 @@
+                self.logger,
+                self.tvu_socket,
+                self.exit,
+            )).channel;
+        }
+        defer for (tvu_receivers) |r| r.deinit();
+        const x = try std.Thread.spawn(
+            .{},
+            Self.runPacketHandler,
+            .{ self, tvu_receivers, sender.channel },
+        );
+        const y = try std.Thread.spawn(
+            .{},
+            Self.runPacketHandler,
+            .{ self, .{repair_receiver.channel}, sender.channel },
+        );
+        x.join();
+        y.join();
+    }
+
+    /// Keep looping over packet channel and process the incoming packets.
+    /// Returns when exit is set to true.
+    fn runPacketHandler(
+        self: *Self,
+        receivers: anytype,
+        sender: *Channel(ArrayList(Packet)),
+    ) !void {
+        var buf = ArrayList(ArrayList(Packet)).init(self.allocator);
+        while (!self.exit.load(.unordered)) {
+            inline for (receivers) |receiver| {
+                var responses = ArrayList(Packet).init(self.allocator);
+                try receiver.tryDrainRecycle(&buf);
+                if (buf.items.len > 0) {
+                    const shred_version = self.shred_version.load(.monotonic);
+                    for (buf.items) |batch| {
+                        for (batch.items) |*packet| {
+                            try self.handlePacket(packet, &responses, shred_version);
+                        }
+                        try self.outgoing_shred_channel.send(batch);
+                    }
+                    if (responses.items.len > 0) {
+                        try sender.send(responses);
+                    }
+                } else {
+                    std.time.sleep(10 * std.time.ns_per_ms);
+                }
+            }
+        }
+    }
+
+    /// Handle a single packet and return
+    fn handlePacket(
+        self: *Self,
+        packet: *Packet,
+        responses: *ArrayList(Packet),
+        shred_version: u16,
+    ) !void {
+        if (packet.size == REPAIR_RESPONSE_SERIALIZED_PING_BYTES) {
+            try self.handlePing(packet, responses);
+            packet.set(.discard);
+        } else {
+            const endpoint_str = try endpointToString(self.allocator, &packet.addr);
+            defer endpoint_str.deinit();
+
+            // TODO set correct values once using snapshot + blockstore
+            const root = 0;
+            const max_slot = std.math.maxInt(Slot);
+            if (shouldDiscardShred(packet, root, shred_version, max_slot)) {
+                packet.set(.discard);
+            }
+        }
+    }
+
+    /// Handle a ping message and return
+    fn handlePing(self: *Self, packet: *const Packet, responses: *ArrayList(Packet)) !void {
+        const repair_ping = bincode.readFromSlice(self.allocator, RepairPing, &packet.data, .{}) catch |e| {
+            self.logger.errf("could not deserialize ping: {} - {any}", .{ e, packet.data[0..packet.size] });
```

Comment:
If a ping message is invalid, we need to stop dealing with that particular message, and move on to processing other messages.

This means we need to skip the rest of the logic in handlePing as soon as we see an issue with the ping. But we shouldn't skip any of the logic outside of handlePing. I felt that the cleanest way to manage that is by having handlePing return early, but not return an error.

It would also be good to have some visibility into the fact that we're receiving invalid messages. That was the goal of the log message. It was logged as `error` but probably should have been `warn` since it doesn't mean there is actually anything going wrong with sig, it just means we've received an invalid message.

The problem with a log message is that anyone who reads this code can find a ddos approach. Send tons of invalid pings to sig, and sig will get bottlenecked in the logger. So I converted it to a metric, which should be less prone to bottlenecks during heavy load.

545e0c6f14e1cb8efedfda0dc08aec2f60fd1353